File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode list = new ListNode();
        ListNode temp1 = list;


        while(list1 != null){
            if(list2 != null && list2.val<list1.val){
                temp1.next = list2;
                list2 = list2.next;
            } else {
                temp1.next = list1;
                list1 = list1.next;
            }
            temp1 = temp1.next;
        }
        
        temp1.next = list2;

        return list.next;
        
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        
        if(not(head.next == None or head.next.next== None)):
            # finding the size of the sll
            size = 0
            current1 = head
            while(current1):
                size+=1
                current1 = current1.next
        
            #resetting current1 to point to head
            current1 = head
            current2 = head
            reqCurrent = current2
            #finding the middle node
            for i in range(0,size//2):
                reqCurrent = current2
                current2 = current2.next
            reqCurrent.next = None
            #reversing the second half of the sll. current2 is the head of this list rn
            prev = None
            temp = None
            while(current2):
                temp = current2.next
                current2.next = prev
                prev = current2
                current2 = temp

            #prev holds the head of the reversed linked list
            temp1 = current1.next
            temp2 = prev.next
            while(temp1 and temp2):
                current1.next = prev
                prev.next = temp1
                prev = temp2
                current1 = temp1
                temp1 = current1.next
                temp2 = prev.next
                
            current1.next = prev

            





        
        

